# __DATA WRANGLING (247): Using personality scores and department information to demonstrate how to manipulate data frames__

# 0. Introduction

This project will focus on working with dataframes using pandas. It will focus on the following key concepts:
* merging data frames
* filtering data frames
* manipulating rows and columns

___

# 1.Initializing Project: Library Setup and Data Cleaning 

## 1.1 Import required libraries


In [21]:
import pandas as pd
import ast
import re


## 1.2 Data cleaning 

In [22]:
personality_df = pd.read_csv("../data/personality_scores.csv", sep=';')

personality_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 70 columns):
 #   Column                                                                    Non-Null Count  Dtype  
---  ------                                                                    --------------  -----  
 0   ID                                                                        1555 non-null   int64  
 1   Section 5 of 6 [I am always prepared.]                                    1555 non-null   object 
 2   Section 5 of 6 [I am easily disturbed.]                                   1555 non-null   object 
 3   Section 5 of 6 [I am exacting (demanding) in my work.]                    1555 non-null   object 
 4   Section 5 of 6 [I am full of ideas.]                                      1555 non-null   object 
 5   Section 5 of 6 [I am interested in people.]                               1555 non-null   object 
 6   Section 5 of 6 [I am not interested in abstract ideas.]         

There appears to be some unnamed columns with only nan values. These will be dropped

In [23]:
personality_df = personality_df.dropna(axis=1)

personality_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 51 columns):
 #   Column                                                                    Non-Null Count  Dtype 
---  ------                                                                    --------------  ----- 
 0   ID                                                                        1555 non-null   int64 
 1   Section 5 of 6 [I am always prepared.]                                    1555 non-null   object
 2   Section 5 of 6 [I am easily disturbed.]                                   1555 non-null   object
 3   Section 5 of 6 [I am exacting (demanding) in my work.]                    1555 non-null   object
 4   Section 5 of 6 [I am full of ideas.]                                      1555 non-null   object
 5   Section 5 of 6 [I am interested in people.]                               1555 non-null   object
 6   Section 5 of 6 [I am not interested in abstract ideas.]                 

After dropping the columns with null values, it would be suitable to provide a summary of the table to get an idea of what the dataframe looks like.

In [24]:
personality_df.head(10)

ID Section 5 of 6 [I am always prepared.]  \
0   0                                 (3, 5)   
1   1                                 (3, 5)   
2   2                                 (3, 5)   
3   3                                 (3, 5)   
4   4                                 (3, 3)   
5   5                                 (3, 5)   
6   6                                 (3, 5)   
7   7                                 (3, 5)   
8   8                                 (3, 5)   
9   9                                 (3, 3)   

  Section 5 of 6 [I am easily disturbed.]  \
0                                  (4, 5)   
1                                  (4, 5)   
2                                  (4, 3)   
3                                  (4, 5)   
4                                  (4, 5)   
5                                  (4, 5)   
6                                  (4, 5)   
7                                  (4, 5)   
8                                  (4, 5)   
9                                  (4, 3)   

  Section 5 of 6 [I am exacting (demanding) in my work.]  \
0                                             (3, 5)       
1                                             (3, 5)       
2                                             (3, 3)       
3                                             (3, 3)       
4                                             (3, 3)       
5                                             (3, 3)       
6                                             (3, 5)       
7                                             (3, 3)       
8                                             (3, 5)       
9                                             (3, 3)       

  Section 5 of 6 [I am full of ideas.]  \
0                               (5, 5)   
1                               (5, 5)   
2                               (5, 5)   
3                               (5, 5)   
4                               (5, 3)   
5                               (5, 5)   
6                               (5, 5)   
7                               (5, 5)   
8                               (5, 5)   
9                               (5, 5)   

  Section 5 of 6 [I am interested in people.]  \
0                                      (2, 3)   
1                                      (2, 5)   
2                                      (2, 5)   
3                                      (2, 5)   
4                                      (2, 3)   
5                                      (2, 5)   
6                                      (2, 5)   
7                                      (2, 5)   
8                                      (2, 5)   
9                                      (2, 5)   

  Section 5 of 6 [I am not interested in abstract ideas.]  \
0                                             (5, 3)        
1                                             (5, 3)        
2                                             (5, 5)        
3                                             (5, 3)        
4                                             (5, 3)        
5                                             (5, 3)        
6                                             (5, 3)        
7                                             (5, 5)        
8                                             (5, 5)        
9                                             (5, 3)        

  Section 5 of 6 [I am not interested in other people's problems.]  \
0                                             (2, 3)                 
1                                             (2, 5)                 
2                                             (2, 5)                 
3                                             (2, 3)                 
4                                             (2, 3)                 
5                                             (2, 5)                 
6                                             (2, 5)                 
7                                             (2, 5)                 
8           

The next part of the cleaning process will be to fix the column names to only have the text in brackets appear.

In [25]:
column_name_pattern = r"\[([^[\]]+)\]"


columns_with_section_in_title = personality_df.columns[1:].str.extract(column_name_pattern).iloc[:, 0]
columns_with_section_in_title = columns_with_section_in_title.str.replace(".", " ")
personality_df.columns = [personality_df.columns[0]] + list(columns_with_section_in_title)
personality_score_df = personality_df

personality_score_df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 51 columns):
 #   Column                                                   Non-Null Count  Dtype 
---  ------                                                   --------------  ----- 
 0   ID                                                       1555 non-null   int64 
 1   I am always prepared                                     1555 non-null   object
 2   I am easily disturbed                                    1555 non-null   object
 3   I am exacting (demanding) in my work                     1555 non-null   object
 4   I am full of ideas                                       1555 non-null   object
 5   I am interested in people                                1555 non-null   object
 6   I am not interested in abstract ideas                    1555 non-null   object
 7   I am not interested in other people's problems           1555 non-null   object
 8   I am not really interested in others  

All the text outside the square parenthesis have been removed and the dataframe has been cleaned enough to begin the process of wrangking the data.

___

# 2. Assignment

* ### Examining the personality scores csv for any duplicates and removing any that exist. 

In [26]:

assert len(personality_score_df) == personality_df.ID.nunique(), "Dataframe lengths do not match"

The code above also includes an assert statement to check that the new data frame is the length of the unique entries of the original data frame.

* ### Creating new columns for the personality test subscales.


To create new columns containing the total score of each personality test subscale, the new columns need to be inserted into the dataframe.

In [27]:
personality_traits = ['Conscientiousness', 'Emotional_Stability', 'Openness', 'Agreeable', 'Extraversion'] 

new_columns = {personality_trait: 0 for personality_trait in personality_traits}

personality_score_totals_df = pd.concat([personality_score_df, pd.DataFrame(new_columns, index=personality_score_df.index)], axis=1)

In [28]:
personality_score_totals_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1555 entries, 0 to 1554
Data columns (total 56 columns):
 #   Column                                                   Non-Null Count  Dtype 
---  ------                                                   --------------  ----- 
 0   ID                                                       1555 non-null   int64 
 1   I am always prepared                                     1555 non-null   object
 2   I am easily disturbed                                    1555 non-null   object
 3   I am exacting (demanding) in my work                     1555 non-null   object
 4   I am full of ideas                                       1555 non-null   object
 5   I am interested in people                                1555 non-null   object
 6   I am not interested in abstract ideas                    1555 non-null   object
 7   I am not interested in other people's problems           1555 non-null   object
 8   I am not really interested in others  

The new columns have been created successfully. However, the "info" attribute indicates that the data type of pre-existing columns, except the ID column, is "object". This type of data cannot be used for performing calculations. Therefore, it is necessary to convert the data type to a more suitable one before calculating.

In [29]:
personality_score_totals_df.iloc[:, :51] = personality_score_totals_df.iloc[:, :51].apply(lambda x: x.map(lambda y: ast.literal_eval(y) if isinstance(y, str) else y))

Now to confirm if the values were converted to tuple

In [30]:
subset_of_data_with_tuple = personality_score_totals_df.iloc[:,1:51]
all_tuples = True 

for col in subset_of_data_with_tuple.columns:
    for row in subset_of_data_with_tuple.index:
        value = subset_of_data_with_tuple.loc[row, col]
        if not isinstance(value, tuple):
            all_tuples = False
            break  

if all_tuples:
    print("All cells are tuples")
else:
    print("Not all cells are tuples")


All cells are tuples


The next step is creating and implementing the function that will add the values of the subscales to their descriptive columns.

In [31]:
def add_to_subscales(row):
    subscale_columns = {1: 'Extraversion', 2: 'Agreeable',3: 'Conscientiousness', 4: 'Emotional_Stability', 5: 'Openness'}
    subscale_totals = {trait: 0 for trait in personality_traits}

    for item_code, response in row[1:51]:
        subscale_code = subscale_columns.get(item_code)
        if subscale_code is not None:
            subscale_totals[subscale_code] += response
    
    return pd.Series(subscale_totals)

total_scores = personality_score_totals_df.apply(add_to_subscales, axis=1) 
personality_scores = personality_score_totals_df.drop(personality_traits, axis=1)

personality_score_totals_df = pd.concat([personality_scores, total_scores], axis=1)

personality_score_totals_df.head()

ID I am always prepared  I am easily disturbed   \
0   0                (3, 5)                 (4, 5)   
1   1                (3, 5)                 (4, 5)   
2   2                (3, 5)                 (4, 3)   
3   3                (3, 5)                 (4, 5)   
4   4                (3, 3)                 (4, 5)   

  I am exacting (demanding) in my work  I am full of ideas   \
0                                (3, 5)              (5, 5)   
1                                (3, 5)              (5, 5)   
2                                (3, 3)              (5, 5)   
3                                (3, 3)              (5, 5)   
4                                (3, 3)              (5, 3)   

  I am interested in people  I am not interested in abstract ideas   \
0                     (2, 3)                                 (5, 3)   
1                     (2, 5)                                 (5, 3)   
2                     (2, 5)                                 (5, 5)   
3                     (2, 5)                                 (5, 3)   
4                     (2, 3)                                 (5, 3)   

  I am not interested in other people's problems   \
0                                          (2, 3)   
1                                          (2, 5)   
2                                          (2, 5)   
3                                          (2, 3)   
4                                          (2, 3)   

  I am not really interested in others  I am quick to understand things   ...  \
0                                (2, 5)                           (5, 5)  ...   
1                                (2, 5)                           (5, 5)  ...   
2                                (2, 5)                           (5, 5)  ...   
3                                (2, 3)                           (5, 3)  ...   
4                                (2, 3)                           (5, 5)  ...   

  I sympathize with others' feelings  I take time out for others   \
0                              (2, 5)                      (2, 5)   
1                              (2, 5)                      (2, 5)   
2                              (2, 5)                      (2, 5)   
3                              (2, 5)                      (2, 5)   
4                              (2, 3)                      (2, 5)   

  I talk to a lot of different people at parties  I use difficult words   \
0                                          (1, 3)                 (5, 1)   
1                                          (1, 5)                 (5, 3)   
2                                          (1, 3)                 (5, 1)   
3                                          (1, 5)                 (5, 1)   
4                                          (1, 3)                 (5, 1)   

  I worry about things  Conscientiousness Emotional_Stability Openness  \
0                (4, 3)                48                  36       42   
1                (4, 3)                46                  40       42   
2                (4, 3)                40                  38       42   
3                (4, 1)                38                  40       38   
4                (4, 3)                46                  38       36   

  Agreeable Extraversion  
0        40           30  
1        46           42  
2        40           28  
3        38           30  
4        34           28  

[5 rows x 56 columns]

The summarized personality score dataframe shows the addition of 5 personality trait columns with the corresponding sum totals for each subscale.

___

* ### Importing the departments.csv into a dataframe. 

The next step is to import the departments.csv file into a dataframe then followed by merging the dataframe with the personality score data frame, keeping all applicants within the department data frame.

The first step is importing the departments.csv dataframe

In [32]:
departments_df = pd.read_csv("../data/departments.csv", sep=';')

departments_df.head()


ID Department
0   0       Data
1   1       Data
2   2       Data
3   3       Data
4   4       Data

The next step is displaying the different departments present.

In [33]:
list_of_departments = departments_df["Department"].unique()
list_of_departments

array(['Data', 'Web Dev', 'Copywriting', 'Design', 'Strategy', 'Web dev'],
      dtype=object)

There appears to be a minor mistake in the departments entered where Web Dev and Web dev were taken as two separate departments. To fix this all occurrences of 'Web dev' will be replaced with 'Web Dev'.

In [34]:
departments_df['Department'] = departments_df['Department'].replace('Web dev', 'Web Dev', regex=True)
list_of_departments = departments_df["Department"].unique()
list_of_departments

array(['Data', 'Web Dev', 'Copywriting', 'Design', 'Strategy'],
      dtype=object)

Now that the issue has been sorted, the two dataframes can be now be merged.

The ‘departments’ dataframe has only 2 columns, ID and Department. The ID column is the only one that is common among the two so merging using the join() function is the best option.

In [35]:
merged_personality_department_df = personality_score_totals_df.merge(departments_df, on='ID', how='inner', suffixes=('_personality_score', '_department'))


assert len(merged_personality_department_df) == len(departments_df), "Dataframe rows are not the same"
expected_num_columns = len(personality_score_totals_df.columns) + len(departments_df.columns) - 1  # Subtract 1 for the overlapping column 'ID'
assert len(merged_personality_department_df.columns) == expected_num_columns, "Number of columns is not as expected"

The assert statement does not return any errors indicating that the dataframes have been successfully merged. Now observe what the merged dataframe looks like.

In [36]:
merged_personality_department_df.head()

ID I am always prepared  I am easily disturbed   \
0   0                (3, 5)                 (4, 5)   
1   1                (3, 5)                 (4, 5)   
2   2                (3, 5)                 (4, 3)   
3   3                (3, 5)                 (4, 5)   
4   4                (3, 3)                 (4, 5)   

  I am exacting (demanding) in my work  I am full of ideas   \
0                                (3, 5)              (5, 5)   
1                                (3, 5)              (5, 5)   
2                                (3, 3)              (5, 5)   
3                                (3, 3)              (5, 5)   
4                                (3, 3)              (5, 3)   

  I am interested in people  I am not interested in abstract ideas   \
0                     (2, 3)                                 (5, 3)   
1                     (2, 5)                                 (5, 3)   
2                     (2, 5)                                 (5, 5)   
3                     (2, 5)                                 (5, 3)   
4                     (2, 3)                                 (5, 3)   

  I am not interested in other people's problems   \
0                                          (2, 3)   
1                                          (2, 5)   
2                                          (2, 5)   
3                                          (2, 3)   
4                                          (2, 3)   

  I am not really interested in others  I am quick to understand things   ...  \
0                                (2, 5)                           (5, 5)  ...   
1                                (2, 5)                           (5, 5)  ...   
2                                (2, 5)                           (5, 5)  ...   
3                                (2, 3)                           (5, 3)  ...   
4                                (2, 3)                           (5, 5)  ...   

  I take time out for others  I talk to a lot of different people at parties   \
0                      (2, 5)                                          (1, 3)   
1                      (2, 5)                                          (1, 5)   
2                      (2, 5)                                          (1, 3)   
3                      (2, 5)                                          (1, 5)   
4                      (2, 5)                                          (1, 3)   

  I use difficult words  I worry about things  Conscientiousness  \
0                 (5, 1)                (4, 3)                48   
1                 (5, 3)                (4, 3)                46   
2                 (5, 1)                (4, 3)                40   
3                 (5, 1)                (4, 1)                38   
4                 (5, 1)                (4, 3)                46   

  Emotional_Stability Openness Agreeable Extraversion Department  
0                  36       42        40           30       Data  
1                  40       42        46           42       Data  
2                  38       42        40           28       Data  
3                  40       38        38           30       Data  
4                  38       36        34           28       Data  

[5 rows x 57 columns]

___

* ### Filtering the data frame by applicants who scored less than 30 on Emotional Stability, Conscientiousness and Agreeableness. 

The first step in this process is defining a function that checks whether the applicant scored less than 30 on Emotional Stability, Conscientiousness and Agreeableness, if this condition is met they are assigned the term "high_risk" in a new column called 'Risk tag' if they do not fall under these conditions they are assigned "low_risk".

In [37]:
def risk_assessment(row):
    high_risk_condition = (
        (row['Emotional_Stability'] < 30) &
        (row['Conscientiousness'] < 30) &
        (row['Agreeable'] < 30))

    return "high_risk" if high_risk_condition else "low_risk"


The next step involves creating a dataframe named "risk_status_df". This dataframe is a filtered version of the "merged_personality_department_df" dataframe that only displays the same results as merged_personality_department_df with the risk tag column as well. To get the risk tag, the risk_assessment function was applied to a column called "risk_tag_column" which displays the risk status from the "merged_personality_department_df" dataframe. The resulting column was merged with the merged_personality_department_df dataframe to create the risk_status_df dataframe. 

In [38]:
risk_tag_column = merged_personality_department_df.apply(risk_assessment, axis=1)
risk_tag_column = risk_tag_column.rename('Risk Status')

risk_status_df =   pd.concat([merged_personality_department_df, risk_tag_column], axis=1)

risk_status_df.head(10)


ID I am always prepared  I am easily disturbed   \
0   0                (3, 5)                 (4, 5)   
1   1                (3, 5)                 (4, 5)   
2   2                (3, 5)                 (4, 3)   
3   3                (3, 5)                 (4, 5)   
4   4                (3, 3)                 (4, 5)   
5   5                (3, 5)                 (4, 5)   
6   6                (3, 5)                 (4, 5)   
7   7                (3, 5)                 (4, 5)   
8   8                (3, 5)                 (4, 5)   
9   9                (3, 3)                 (4, 3)   

  I am exacting (demanding) in my work  I am full of ideas   \
0                                (3, 5)              (5, 5)   
1                                (3, 5)              (5, 5)   
2                                (3, 3)              (5, 5)   
3                                (3, 3)              (5, 5)   
4                                (3, 3)              (5, 3)   
5                                (3, 3)              (5, 5)   
6                                (3, 5)              (5, 5)   
7                                (3, 3)              (5, 5)   
8                                (3, 5)              (5, 5)   
9                                (3, 3)              (5, 5)   

  I am interested in people  I am not interested in abstract ideas   \
0                     (2, 3)                                 (5, 3)   
1                     (2, 5)                                 (5, 3)   
2                     (2, 5)                                 (5, 5)   
3                     (2, 5)                                 (5, 3)   
4                     (2, 3)                                 (5, 3)   
5                     (2, 5)                                 (5, 3)   
6                     (2, 5)                                 (5, 3)   
7                     (2, 5)                                 (5, 5)   
8                     (2, 5)                                 (5, 5)   
9                     (2, 5)                                 (5, 3)   

  I am not interested in other people's problems   \
0                                          (2, 3)   
1                                          (2, 5)   
2                                          (2, 5)   
3                                          (2, 3)   
4                                          (2, 3)   
5                                          (2, 5)   
6                                          (2, 5)   
7                                          (2, 5)   
8                                          (2, 5)   
9                                          (2, 5)   

  I am not really interested in others  I am quick to understand things   ...  \
0                                (2, 5)                           (5, 5)  ...   
1                                (2, 5)                           (5, 5)  ...   
2                                (2, 5)                           (5, 5)  ...   
3                                (2, 3)                           (5, 3)  ...   
4                                (2, 3)                           (5, 5)  ...   
5                                (2, 5)                           (5, 5)  ...   
6                                (2, 5)                           (5, 5)  ...   
7                                (2, 5)                           (5, 5)  ...   
8                                (2, 5)                           (5, 5)  ...   
9                                (2, 3)                           (5, 5)  ...   

  I talk to a lot of different people at parties  I use difficult words   \
0                                          (1, 3)                 (5, 1)   
1                                          (1, 5)                 (5, 3)   
2                                          (1, 3)                 (5, 1)   
3                                          (1, 5)                 (5, 1)   
4                                          (1, 3)                 (5, 1)   


The next step is to identitfy the high risk applicants from this dataframe.

In [39]:
high_risk_applicants_df = risk_status_df[risk_status_df["Risk Status"] == "high_risk"][["ID", "Department"]]

high_risk_applicants_df

ID   Department
881    881         Data
1197  1197  Copywriting

The results of the filter show that only two applicants, applicant no. 881 in the data department and applicant no. 1197 in the copywriting department, are classified as 'high risk'.

___

* ### Creating a new data frame with a count of the number of low and high-risk applicants within each department.

In [40]:
risk_status_summary_df = risk_status_df.groupby(['Risk Status', 'Department']).size().unstack(fill_value=0)
risk_status_summary_df.index.name = None
risk_status_summary_df.columns.name = 'Risk Status'

risk_status_summary_df = risk_status_summary_df.rename(index={'low_risk': 'Low risk', 'high_risk': 'High risk'})

risk_status_summary_df


Risk Status  Copywriting  Data  Design  Strategy  Web Dev
High risk              1     1       0         0        0
Low risk             325   328     120       449      331

___

# 3. Conclusion

The aim of this project was to showcase the different data wrangling skills by applying functions to a dataframe, filtering, sorting, and joining different dataframes together. This was achieved this by completing the exercises throughout this [page](http://syllabus.africacode.net/projects/data-science-specific/data-wrangling/).